src/NServiceBus.CustomChecks/TimerBasedPeriodicCheck.cs

```diff
@@ -1,105 +1,121 @@
-namespace NServiceBus.CustomChecks
+namespace NServiceBus.CustomChecks;
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Logging;
+using ServiceControl.Plugin.CustomChecks.Messages;
+
+sealed class TimerBasedPeriodicCheck(
+    ICustomCheckWrapper check,
+    ServiceControlBackend messageSender,
+    Func<CustomCheckHostInfo, ICustomCheck, CheckResult, ReportCustomCheckResult> messageFactory,
+    TimeSpan ttl,
+    CustomCheckHostInfo customCheckHostInfo) : IAsyncDisposable
 {
-    using System;
-    using System.Threading;
-    using System.Threading.Tasks;
-    using Logging;
-    using ServiceControl.Plugin.CustomChecks.Messages;
-
-    sealed class TimerBasedPeriodicCheck(
-        ICustomCheckWrapper check,
-        ServiceControlBackend messageSender,
-        Func<CheckResult, ReportCustomCheckResult> messageFactory,
-        TimeSpan ttl) : IAsyncDisposable
+    static readonly ILog Logger = LogManager.GetLogger<TimerBasedPeriodicCheck>();
+
+    CancellationTokenSource? stopTokenSource;
+    Task? timerTask;
+
+    public void Start()
     {
-        static readonly ILog Logger = LogManager.GetLogger<TimerBasedPeriodicCheck>();
+        stopTokenSource = new CancellationTokenSource();
 
-        CancellationTokenSource stopTokenSource;
+        timerTask = RunAndSwallowExceptions(stopTokenSource.Token);
+    }
 
-        public void Start()
+    public async Task Stop(CancellationToken cancellationToken = default)
+    {
+        if (stopTokenSource is not null)
         {
-            stopTokenSource = new CancellationTokenSource();
-
-            _ = RunAndSwallowExceptions(stopTokenSource.Token);
+            await stopTokenSource.CancelAsync().ConfigureAwait(false);
         }
 
-        public Task Stop(CancellationToken cancellationToken = default)
+        if (timerTask is null)
         {
-            stopTokenSource?.Cancel();
-
-            return Task.CompletedTask;
+            return;
         }
 
-        public ValueTask DisposeAsync()
+        try
         {
-            stopTokenSource?.Dispose();
-            return check.DisposeAsync();
+            await timerTask.WaitAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
I thought it is cleaner to await the outcome for the timer task to properly exit given that we support proper cancellation and the host stop token is passed through. Arguably, though, this is a change that might make the stop take in some cases slightly longer but given we also support disposing now, I think it makes sense.